rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/TreeView.java

```diff
@@ -271,150 +270,127 @@ private void enableItems(MenuItem[] items) {
-     * Map)
-     */
-    public void nodesUpdated(Map<String, NodeSource> nodes) {
-
-        /*
-         * Add to _this.tree_ the nodes contained in _nodes_ but not present
-         * in _oldNodes_
-         */
-        for (NodeSource ns : nodes.values()) {
-            String nsName = ns.getSourceName();
-
-            /* NodeSources */
-            NodeSource oldNs = (oldNodes != null) ? oldNodes.get(nsName) : null;
-            if (oldNs == null) {
-                /* new node source */
-                TNS nsTreeNode = new TNS(getNodeSourceDisplayedDescription(ns, nsName), ns);
-                nsTreeNode.setAttribute("nodeId", nsName);
-                nsTreeNode.setIcon(ns.getIcon());
-                this.tree.add(nsTreeNode, this.tree.getRoot());
-                this.curNodes.put(nsName, nsTreeNode);
-            } else {
-                /* node source update */
-                if (!oldNs.getNodeSourceStatus().equals(ns.getNodeSourceStatus()) ||
-                    !oldNs.getSourceDescription().equals(ns.getSourceDescription())) {
-                    TNS curTreeNodeSource = (TNS) curNodes.get(nsName);
-                    curTreeNodeSource.rmNS = ns;
-                    curTreeNodeSource.setName(getNodeSourceDisplayedDescription(ns, nsName));
-                    curTreeNodeSource.setIcon(ns.getIcon());
-                    this.treeGrid.refreshFields();
-                }
-            }
+    @Override
+    public void updateByDelta(List<NodeSource> nodeSources, List<Node> nodes) {
+        processNodeSources(nodeSources);
 
-            for (Node n : ns.getDeploying().values()) {
-                String nodeUrl = n.getNodeUrl();
-                TNode nodeTreeNode = new TNode(nodeUrl, n);
-                nodeTreeNode.setAttribute("nodeId", nodeUrl);
-
-                /* Deploying nodes */
-                Node oldNode = (oldNs != null) ? oldNs.getDeploying().get(nodeUrl) : null;
-                if (oldNode == null) {
-                    this.tree.add(nodeTreeNode, this.curNodes.get(nsName));
-                    this.curNodes.put(nodeUrl, nodeTreeNode);
-                    nodeTreeNode.setIcon(n.getIcon());
-                } else {
-                    this.curNodes.get(nodeUrl).setIcon(n.getIcon());
-                }
-            }
+        treeGrid.refreshFields();
 
-            for (Host h : ns.getHosts().values()) {
-                String hostName = h.getHostName();
-                THost hostTreeNode = new THost(hostName, h);
-                hostTreeNode.setAttribute("nodeId", h.getId());
-                if (h.isVirtual()) {
-                    hostTreeNode.setIcon(RMImages.instance.host_virtual_16().getSafeUri().asString());
-                } else {
-                    hostTreeNode.setIcon(RMImages.instance.host_16().getSafeUri().asString());
-                }
+        processNodes(nodes);
+
+        treeGrid.markForRedraw();
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 75 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)